crates/task-impls/src/quorum_vote/drb_computations.rs

```diff
@@ -0,0 +1,158 @@
+use std::collections::{btree_map, BTreeMap};
+
+use hotshot_types::{
+    drb::{compute_drb_result, DrbResult, DrbSeedInput},
+    traits::node_implementation::{ConsensusTime, NodeType},
+};
+use tokio::{spawn, task::JoinHandle};
+
+/// Number of previous results and seeds to keep
+pub const KEEP_PREVIOUS_RESULT_COUNT: u64 = 8;
+
+/// Helper struct to track state of DRB computations
+pub struct DrbComputations<TYPES: NodeType> {
+    /// Stored results from computations
+    results: BTreeMap<TYPES::Epoch, DrbResult>,
+
+    /// Currently live computation
+    task: Option<(TYPES::Epoch, JoinHandle<DrbResult>)>,
+
+    /// Stored inputs to computations
+    seeds: BTreeMap<TYPES::Epoch, DrbSeedInput>,
+}
+
+impl<TYPES: NodeType> DrbComputations<TYPES> {
+    #[must_use]
+    /// Create a new DrbComputations
+    pub fn new() -> Self {
+        Self {
+            results: BTreeMap::new(),
+            task: None,
+            seeds: BTreeMap::new(),
+        }
+    }
+
+    /// If a task is currently live AND has finished, join it and save the result.
+    /// If a task is currently live and NOT finished, abort it.
+    /// self.task will always be None after this call.
+    async fn join_or_abort_task(&mut self) {
+        if let Some((epoch, join_handle)) = &mut self.task {
+            if join_handle.is_finished() {
+                match join_handle.await {
+                    Ok(result) => {
+                        self.results.insert(*epoch, result);
+                    }
+                    Err(e) => {
+                        tracing::error!("error joining DRB computation task: {e:?}");
+                    }
+                }
+            } else {
+                join_handle.abort();
+            }
+            self.task = None;
+        }
+    }
+
+    /// If a task is currently live AND has finished, join it and save the result.
+    /// If the epoch for the calculation was the same as the provided epoch, return true
+    /// If a task is currently live and NOT finished, abort it UNLESS the task epoch is the same as
+    /// cur_epoch, in which case keep letting it run and return true.
+    /// Return false if a task should be spawned for the given epoch.
+    async fn join_or_abort_task_for_epoch(&mut self, epoch: TYPES::Epoch) -> bool {
```

Comment:
Nit: Probably better to call it `join_or_abort_old_task_for_epoch` or `join_or_abort_old_task` to make it explicit that we'll only abort a task if it's not the same as the given epoch.